esileapclient/osc/v1/mdc/mdc_offer.py

```diff
@@ -0,0 +1,216 @@
+            'resource_type': parsed_args.resource_type,
+        }
+
+        for c in cloud_regions:
+            client = esileapclient.Client(session=c.get_session())
+
+            offers = client.offer.list(filters)
+            for offer in offers:
+                offer.cloud = c.name
+                offer.region = c.config['region_name']
+                data += [offer]
+
+        if parsed_args.long:
+            columns = OFFER_RESOURCE.detailed_fields.keys()
+            labels = OFFER_RESOURCE.detailed_fields.values()
+        else:
+            columns = OFFER_RESOURCE.fields.keys()
+            labels = OFFER_RESOURCE.fields.values()
+        columns = ['cloud', 'region'] + list(columns)
+        labels = ['Cloud', 'Region'] + list(labels)
+
+        return (labels,
+                (oscutils.get_item_properties(s, columns) for s in data))
+
+
+class MDCClaimOffer(command.Lister):
+    """Claim offers across multiple data centers."""
+
+    log = logging.getLogger(__name__ + ".MDCClaimOffer")
+    auth_required = False
+
+    def get_parser(self, prog_name):
+        parser = super(MDCClaimOffer, self).get_parser(prog_name)
+
+        parser.add_argument(
+            "node_count",
+            metavar="<node_count>",
+            help="Number of nodes to claim")
+        parser.add_argument(
+            'start_time',
+            metavar='<start_time>',
+            help="Time when the lease will start.")
+        parser.add_argument(
+            'end_time',
+            metavar='<end_time>',
+            help="Time when the lease will expire.")
+        parser.add_argument(
+            '--clouds',
+            dest='clouds',
+            metavar='<clouds>',
+            nargs="+",
+            help="Specify the cloud to use from clouds.yaml."
+        )
+        parser.add_argument(
+            '--resource-type',
+            dest='resource_type',
+            required=False,
+            help="Show all offers with given resource-type.")
+
+        return parser
+
+    def take_action(self, parsed_args):
+        cloud_regions = openstack.config.loader.OpenStackConfig().\
+            get_all_clouds()
+        if parsed_args.clouds:
+            cloud_regions = filter(lambda c: c.name in parsed_args.clouds,
+                                   cloud_regions)
+        node_count = int(parsed_args.node_count)
+        filters = {
+            'status': 'available',
+            'available_start_time': str(parsed_args.start_time) if
+            parsed_args.start_time else None,
+            'available_end_time': str(parsed_args.end_time) if
+            parsed_args.end_time else None,
+            'resource_type': parsed_args.resource_type,
+        }
+
+        available_offers = []
+        for c in cloud_regions:
+            client = esileapclient.Client(session=c.get_session())
+            offers = client.offer.list(filters)
+            for offer in offers:
+                offer.cloud_region = c
```

Comment:
why is this being set? I see we use it to get a session, but why do we need individual sessions for every offer if multiple of the offers could all be within one of the datacenters

Reply:
Right now we just grab X random offers, which come out in any order. I agree that in the long run there's probably a better algorithm, and people will probably prefer nodes to be constrained to one data center. But I'd like to wait for outside requirements before doing anything more complicated.